### PR TITLE
Feature kd global

### PR DESCRIPTION
**Pull request by beykyle**
The main change here is in `koning_delaroche.py`, where a class for the global KD OMP has been implemented.
- [x] #48 

This also addresses:

- [x] #62 
- [x] #60 
- [x] #55
- [x]  #51
- [x] #50
- [x] #42 
- [x] #41 
- [x] #40 
- [x] #39 
- [x] #10 - although not in the way originally proposed in the issue. Instead, the partial wave series is dynamically truncated based on convergence of relative difference between scattering amplitude magnitudes. To my knowledge, this is the more standard way to handle this problem. The relative tolerance for this cutoff is a user parameter.
- [x] #13 - which I would argue was already implemented in 0.9.1, unless I'm misunderstanding

I also added a a "tutorial 0", or quickstart, which has the minimal code and description to emulate a simple problem. I argue this is nice to have as (at least for me), this is the kind of thing I look for first when I go through code documentation. CAT-analysis and UQ are more advanced topics, I'd argue. Along with this, I will attempt to simplify the existing tutorials by factoring some of the repeated code into convenience functions, so that they primarily focus on the main ideas. Very open to feedback about this (and in general) -  I don't want to remove so much that clarity is impacted.
- [x] #46

Also, because I have Numerov implemented for another code, I'll throw this in for free:
- [x] #24 

While I'm at it, because I love procrastinating on paper writing with writing code:

- [x] #11

Finally, one other thing I experimented with is the use of the `numba.njit` to compile the calculation of cross sections in the code. It is a bit finicky, but shows potential for speedup if other kernels in the code can be simplified as well.

### CODE REVIEW

**Comment by beykyle** [comment]
Also, apologies for the extra lines in the diffs. My editor automatically removes extra whitespace. I tried to avoid committing most of the resulting changes, but a few slipped through. I can go through and remove them.

On the other hand, I would argue we should just standardize formatting across the code to minimize diff sizes for easier reviews. I like `[black](https://pypi.org/project/black/)` personally for its simplicity. 

I can make a separate PR that formats the existing code, and adds a workflow to format code being pulled into develop for easier review.

**Comment by kylegodbey** [comment]
Any thoughts on JAX? I don't think we can necessarily take advantage of GPUs for current workloads, but it is nice you get it for free with the JAX jit and CPU parallelism

**Comment by beykyle** [comment]
@kylegodbey I haven't tried JAX, but have heard good things. I'm trying to use [numba.njit](https://numba.pydata.org/numba-doc/latest/glossary.html#term-nopython-mode), which seems promising, but is somewhat finicky when integrating in object oriented code. It ends up making the code less  concise and pythonic, so if doesn't end up paying off in performance, I'll take it out and try JAX or Cython (or give it up and let the next ambitious grad student figure it out).

**Comment by beykyle** [comment]
Just updating this to include some of the miscellaneous things that are included in  this branch:
- least-squares as an alternative to collocation for EIM
- ability to "freeze" parameters by providing equal upper and lower bounds in training info. This is useful for global OM emulation because, for example, if one emulates just along energy, some parameters will be frozen and some wont. Now this is handled automatically.
- `percent_explained_variance` metric in `basis` and `InteractionEIM`
- General refactoring of inheritance to minimize repeated code

**Comment by kylegodbey** [comment]
Is this feature complete at this point? Now that the rose paper is nearing publication I think we can move towards our first big feature update

**Comment by beykyle** [comment]
> Is this feature complete at this point? Now that the rose paper is nearing publication I think we can move towards our first big feature update

Not quite, the last piece is a set of notebooks developing an example for emulating a global OMP. Once I graduate and get over to EL, we can have a discussion on if/how we get that integrated into the code. I wont have time to really clean things until then.

If you’re in the mood for a code review, however, that’s always helpful!

**Comment by beykyle** [comment]
This does fix a lot of things compared to the current release, so we may want to consider pulling this into develop if there are people itching to add new features in the interim.

**Comment by beykyle** [comment]
@kylegodbey @pabggpnMSU I am officially requesting a review on this branch. I know it's a huge change (I'm sorry).

**Comment by kylegodbey** [comment]
> @kylegodbey @pabggpnMSU I am officially requesting a review on this branch. I know it's a huge change (I'm sorry).

Yee-haw, I'll try to look this weekend. Did you add any new tests for the feature updates? I'll definitely pull this into another branch just in case it breaks our build workflows.

**Comment by beykyle** [comment]
No new tests are added, but I can def write some. Old tests still pass, and the tutorials are updated and still run properly (better in fact, as many bugs are fixed; the issues causing the CAT plot elbows for example).

The build workflows have not been pulled into main yet, in fact they are introduced in this branch. See https://github.com/bandframework/rose/actions/workflows/python-package.yml

This workflow introduces testing, linting, and automatic formatting.

**Comment by beykyle** [comment]
For future reference, attached are some profiler results from the emulator constructed in [tutorial 0](https://github.com/bandframework/rose/blob/2f89c7bdac0c932b8d1a38d663bd5cbdc0f679f2/docs/tutorials/ROSE_tutorial_0_quickstart.ipynb).

For the high fidelity:
```
%%prun
emulator.exact_xs(test[0])
```
```
 4133008 function calls in 3.644 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    10773    0.640    0.000    3.085    0.000 rk.py:14(rk_step)
    64696    0.457    0.000    2.316    0.000 schroedinger.py:155(<lambda>)
   258784    0.330    0.000    0.668    0.000 functions.py:524(__init__)
   258784    0.244    0.000    1.559    0.000 abstract.py:61(__call__)
   258784    0.222    0.000    0.222    0.000 dispatcher.py:721(__repr__)
    10712    0.177    0.000    3.461    0.000 rk.py:111(_step_impl)
...
```
and so on.

For the (40,15) emulator below:
```
%%prun
emulators[4].emulate_xs(test[0])
```

```
         1727 function calls in 0.018 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       29    0.007    0.000    0.007    0.000 {built-in method numpy.core._multiarray_umath.c_einsum}
       29    0.003    0.000    0.004    0.000 linalg.py:329(solve)
       29    0.002    0.000    0.016    0.001 reduced_basis_emulator.py:213(coefficients)
       29    0.001    0.000    0.002    0.000 interaction.py:96(tilde)
       29    0.001    0.000    0.018    0.001 reduced_basis_emulator.py:258(R_matrix_element)
```

For the (3,15) emulator:

```
         1727 function calls in 0.013 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       29    0.004    0.000    0.013    0.000 reduced_basis_emulator.py:258(R_matrix_element)
       29    0.002    0.000    0.007    0.000 reduced_basis_emulator.py:213(coefficients)
       29    0.001    0.000    0.002    0.000 interaction.py:96(tilde)
       29    0.001    0.000    0.002    0.000 linalg.py:329(solve)
       29    0.001    0.000    0.001    0.000 shape_base.py:292(hstack)
       29    0.001    0.000    0.003    0.000 interaction_eim.py:150(coefficients)
       29    0.001    0.000    0.001    0.000 {built-in method numpy.core._multiarray_umath.c_einsum}
```

Here is the CAT plot that includes these configurations, which shows both accuracy and performance ceilings:
cat_emu

The upshot is that, for (20,15) and (40,15), we are spending most of our time 1) calculating the RBM matrix  2) solving it. For the (3,15) and (5,15), we are spending a greater fraction of our time in pythony things that would benefit from the online toolchain being a compiled c++ extension. Based on this profiling, I would recommend that this extension be built.

**Review by kylegodbey — APPROVED** [review]
All looks good to me, I had a few thoughts and questions throughout but I am good to merge.

File: src/rose/interaction.py [hunk 1]
```diff
@@ -1,28 +1,32 @@
...
-'''
+"""
 
 from typing import Callable
 import numpy as np
+from numba import njit
+
 from .constants import HBARC, ALPHA
-from .spin_orbit import SpinOrbitTerm
+from .spin_orbit import SpinOrbitTerm, null
```
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/interaction.py:10]: This is a bit hacky admittedly, I'll remove the need for it

File: src/rose/interaction_eim.py [hunk 2]
```diff
@@ -110,159 +71,166 @@ def __init__(self,
...
+        """
+        assert training_info is not None
+
+        if n_basis is None:
+            if "n_theta" in kwargs:
+                n_basis = kwargs["n_theta"]
+            else:
+                n_basis = 8
+
+        super().__init__(**kwargs)
```
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/interaction_eim.py:83]: No great reason, just seem sensible.

File: src/rose/interaction_eim.py [hunk 3]
```diff
@@ -110,159 +71,166 @@ def __init__(self,
...
+            self.match_indices = np.array(
+                [np.argmin(np.abs(rho_mesh - ri)) for ri in self.match_points]
+            )
+            self.r_i = rho_mesh[self.match_indices]
+            self.Ainv = np.linalg.inv(self.snapshots[self.match_indices])
+        elif match_points is None and method == "collocation":
             # random r points between 0 and 2π fm
-            i_max = self.snapshots.shape[0] // 2
-            di = i_max // (n_basis - 1)
+            i_max = self.snapshots.shape[0] // 4
```
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/interaction_eim.py:127]: This is not super robust. The idea is to seed the MAXVOL with points where the potential varies the most, which happens typically in the first 1/4th of the radial domain. A future commit should improve this but works well for now.

[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/interaction_eim.py:127]: This is not super robust. The idea is to seed the MAXVOL with points where the potential varies the most, which happens typically in the first 1/4th of the radial domain. A future commit should improve this but works well for now.

File: src/rose/interaction_eim.py [hunk 4]
```diff
@@ -110,159 +71,166 @@ def __init__(self,
...
+
         Returns:
             u_j (ndarray): "pillars" (MxN matrix; M = number of mesh points; N = number of pillars)
 
-        '''
+        """
         return np.copy(self.snapshots)
-    
+
+    def percent_explained_variance(self, n=None):
```
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/interaction_eim.py:189]: yeah, it's a bit different for an RBM basis than an EIM basis

File: src/rose/lagrangelegendersolver.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
+from .schroedinger import SchroedingerEquation
+from .interaction import Interaction
+from .constants import HBARC
+
+import numpy as np
+import jitr
+
+class LagrangeRmatrix(SchroedingerEquation):
```
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/lagrangelegendersolver.py:9]: I use it, but could see the argument for not including it in ROSE

File: src/rose/njit_solver_utils.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    return -1 * (
+        (v_r(s / k, alpha) + v_so(s / k, alpha, l_dot_s)) / E
+        + 2 * eta * regular_inverse_s(s, S_C)
+        + l * (l + 1) / s**2
+        - 1.0
+    )
+
+
+@njit
+def numerov_kernel(
```
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/njit_solver_utils.py:55]: Yes, I have that in `docs/tutorials/ROSE_Add_your_own_HF_solver.ipynb`

File: src/rose/wlh.py
[the remark this answers is not in the document]
Comment (beykyle) [reply src/rose/wlh.py:1]: I will add unit tests

**Pull request merged**